eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentPacket.scala

```diff
@@ -42,40 +42,68 @@ object IncomingPaymentPacket {
+    val amountToForward: MilliSatoshi = payload.amountToForward(add.amountMsat)
+    val outgoingCltv: CltvExpiry = payload.outgoingCltv(add.cltvExpiry)
+    val relayFeeMsat: MilliSatoshi = add.amountMsat - amountToForward
+    val expiryDelta: CltvExpiryDelta = add.cltvExpiry - outgoingCltv
   }
   /** We must relay the payment to a remote node. */
-  case class NodeRelayPacket(add: UpdateAddHtlc, outerPayload: PaymentOnion.FinalTlvPayload, innerPayload: PaymentOnion.NodeRelayPayload, nextPacket: OnionRoutingPacket) extends RelayPacket
+  case class NodeRelayPacket(add: UpdateAddHtlc, outerPayload: PaymentOnion.FinalPayload, innerPayload: PaymentOnion.NodeRelayPayload, nextPacket: OnionRoutingPacket) extends RelayPacket
   // @formatter:on
 
-  case class DecodedOnionPacket[T <: PaymentOnion.PacketType](payload: T, next: OnionRoutingPacket)
+  case class DecodedOnionPacket(payload: TlvStream[OnionPaymentPayloadTlv], next_opt: Option[OnionRoutingPacket])
 
-  private[payment] def decryptOnion[T <: PaymentOnion.PacketType](paymentHash: ByteVector32, privateKey: PrivateKey, packet: OnionRoutingPacket, perHopPayloadCodec: Boolean => Codec[T])(implicit log: LoggingAdapter): Either[FailureMessage, DecodedOnionPacket[T]] =
+  private[payment] def decryptOnion(paymentHash: ByteVector32, privateKey: PrivateKey, packet: OnionRoutingPacket): Either[FailureMessage, DecodedOnionPacket] =
     Sphinx.peel(privateKey, Some(paymentHash), packet) match {
-      case Right(p@Sphinx.DecryptedPacket(payload, nextPacket, _)) =>
-        perHopPayloadCodec(p.isLastPacket).decode(payload.bits) match {
-          case Attempt.Successful(DecodeResult(perHopPayload, _)) => Right(DecodedOnionPacket(perHopPayload, nextPacket))
-          case Attempt.Failure(e: OnionRoutingCodecs.MissingRequiredTlv) => Left(e.failureMessage)
-          case Attempt.Failure(e: OnionRoutingCodecs.ForbiddenTlv) => Left(e.failureMessage)
-          // Onion is correctly encrypted but the content of the per-hop payload couldn't be parsed.
-          // It's hard to provide tag and offset information from scodec failures, so we currently don't do it.
-          case Attempt.Failure(_) => Left(InvalidOnionPayload(UInt64(0), 0))
+      case Right(p: Sphinx.DecryptedPacket) =>
+        PaymentOnionCodecs.perHopPayloadCodec.decode(p.payload.bits) match {
+          case Attempt.Successful(DecodeResult(perHopPayload, _)) if p.isLastPacket => Right(DecodedOnionPacket(perHopPayload, None))
+          case Attempt.Successful(DecodeResult(perHopPayload, _)) => Right(DecodedOnionPacket(perHopPayload, Some(p.nextPacket)))
+          case Attempt.Failure(_) =>
+            // Onion is correctly encrypted but the content of the per-hop payload couldn't be decoded.
+            // It's hard to provide tag and offset information from scodec failures, so we currently don't do it.
+            Left(InvalidOnionPayload(UInt64(0), 0))
```

Comment:
Logging the decrypted bytes could be really helpful for debugging.

Reply:
I wasn't a big fan of adding a logger to this function, and it's also somewhat of a DoS vector if we logged those. This error happens when the sender is buggy, I'd rather have them log their data to figure out what they did wrong.

We could also get spammed by such errors that could force us to log 1kB every time, which isn't great.